packages/execute-driver/lib/plugin.js

```diff
@@ -0,0 +1,134 @@
+      throw new Error(`Only the '${SCRIPT_TYPE_WDIO}' script type is currently supported`);
+    }
+
+    if (!driver.serverHost || !driver.serverPort) {
+      throw new Error('Address or port of running server were not defined; this ' +
+                      'is required. This is probably a programming error in the driver');
+    }
+
+    if (!_.isNumber(timeout)) {
+      throw new Error('Timeout parameter must be a number');
+    }
+
+    const driverOpts = {
+      sessionId: driver.sessionId,
+      protocol: 'http', // Appium won't ever be behind ssl locally
+      hostname: driver.serverHost,
+      port: driver.serverPort,
+      path: driver.serverPath,
+      isW3C: true,
+      isMobile: true,
+      capabilities: driver.caps
+    };
+    this.logger.info(`Constructed webdriverio driver options; W3C mode is ${driverOpts.isW3C ? 'on' : 'off'}`);
+
+
+    // fork the execution script as a child process
+    const childScript = path.join(__dirname, 'execute-child.js');
+    this.logger.info(`Forking process to run webdriver script as child using ${childScript}`);
+    const scriptProc = cp.fork(childScript);
+
+    // keep track of whether we have canceled the script timeout, so we can stop
+    // waiting for it and allow this process to finish gracefully
+    let timeoutCanceled = false;
+
+    try {
+      const timeoutStart = Date.now();
+
+      // promise that deals with the result from the child process
+      const waitForResult = async () => {
+        const resPromise = new B((res) => {
+          scriptProc.on('message', res); // this is node IPC
+        });
+
+        const res = await resPromise;
+        this.logger.info('Received execute driver script result from child process, shutting it down');
+
+        if (res.error) {
+          throw new Error(res.error.message);
+        }
+
+        return res.success;
+      };
+
+      // promise that waits up to the timeout and throws an error if so, or does
+      // nothing if the timeout is canceled because we got a result from the
+      // child script
+      const waitForTimeout = async function () {
+        while (!timeoutCanceled && (Date.now() - timeoutStart) < timeout) {
+          await B.delay(500);
+        }
+
+        if (timeoutCanceled) {
+          return;
+        }
+
+        throw new Error(`Execute driver script timed out after ${timeout}ms. ` +
+                        `You can adjust this with the 'timeout' parameter.`);
+      };
+
+      // now that the child script is alive, send it the data it needs to start
+      // running the driver script
+      this.logger.info('Sending driver and script data to child');
+      scriptProc.send({driverOpts, script, timeout});
+
+      // and set up a race between the response from the child and the timeout
+      return await B.race([waitForResult(), waitForTimeout()]);
+    } catch (err) {
+      throw new Error(`Could not execute driver script. Original error was: ${err}`);
```

Comment:
FWIW, https://github.com/tc39/proposal-error-cause is now part of the standard

Reply:
the user never has a chance to catch this error object though, so `err.cause` will be useless. it's flattened into the W3C response JSON and shown in the appium logs. so i think the best we can do is be good in logging it for debug purposes.